smithy-model/src/main/java/software/amazon/smithy/model/validation/validators/HttpQueryParamsTraitValidator.java

```diff
@@ -0,0 +1,82 @@
+import software.amazon.smithy.model.traits.HttpQueryParamsTrait;
+import software.amazon.smithy.model.traits.HttpQueryTrait;
+import software.amazon.smithy.model.traits.Trait;
+import software.amazon.smithy.model.validation.AbstractValidator;
+import software.amazon.smithy.model.validation.ValidationEvent;
+
+/**
+ * When the `httpQueryParams` trait is used, this validator emits a NOTE when another member of the container shape
+ * applies the `httpQuery` trait which may result in a conflict within the query string.
+ */
+public final class HttpQueryParamsTraitValidator extends AbstractValidator {
+    @Override
+    public List<ValidationEvent> validate(Model model) {
+        if (!model.isTraitApplied(HttpQueryParamsTrait.class)) {
+            return Collections.emptyList();
+        } else {
+            return validateQueryTraitUsage(model);
+        }
+    }
+
+    private List<ValidationEvent> validateQueryTraitUsage(Model model) {
+        List<ValidationEvent> events = new ArrayList<>();
+
+        for (Shape shape : model.getShapesWithTrait(HttpQueryParamsTrait.class)) {
+            shape.asMemberShape().flatMap(member -> model.getShape(member.getContainer())
+                .flatMap(Shape::asStructureShape))
+                .ifPresent(structure -> {
+                    // Gather the names of member shapes, as strings, that apply HttpQuery or HttpQueryParams traits
+                    List<String> queryShapes = getMembersWithTrait(structure, HttpQueryTrait.class);
+                    List<String> queryParamsShapes = getMembersWithTrait(structure, HttpQueryParamsTrait.class);
+                    // Only one HttpQueryParams trait can be applied per structure.  If it isn't applied, no note will
+                    // be emitted.
+                    if (queryShapes.size() > 0 && queryParamsShapes.size() == 1) {
+                        events.add(createNote(structure, queryParamsShapes.get(0), queryShapes));
+                    }
+                });
+        }
+
+        return events;
+    }
+
+    private List<String> getMembersWithTrait(StructureShape structure, Class<? extends Trait> trait) {
+        List<String> members = new ArrayList<>();
+        for (MemberShape member : structure.members()) {
+            if (member.hasTrait(trait)) {
+                members.add(member.toShapeId().toString());
+            }
+        }
+        return members;
+    }
+
+    private ValidationEvent createNote(Shape target, String queryParamsShape, List<String> queryShapes) {
+        return note(target, String.format("Operation input `%s` has an `httpQueryParams` trait applied to the `%s` "
+                + "member and `httpQuery` traits applied to the following members: %s. This can cause confusion when "
+                + "keys from the `httpQueryParams` trait conflict with those defined directly by `httpQuery` traits",
+                target.toShapeId(), queryParamsShape, String.join(", ", queryShapes)));
```

Comment:
nit: use [`ValidationUtils.tickedList`](https://github.com/awslabs/smithy/blob/dc40d2beee3b3b97c2349e64afbd68214c72187b/smithy-model/src/main/java/software/amazon/smithy/model/validation/ValidationUtils.java#L76) for the `queryShapes` list.

Reply:
Fixed in 7004468.